feat(js): Update custom trace propagation docs (#12475)
getsentry/sentry-docs: platform-includes/distributed-tracing/custom-instrumentation/browser/javascript.mdx

Distributed tracing will be set up automatically if you've <PlatformLink to="/tracing/">set up tracing</PlatformLink>.

When tracing is enabled, the Sentry SDK will not only create spans for pageloads and transactions but also propagate trace information to other systems. This allows you to connect multiple systems and see how they interact with each other.

If you have not enabled tracing, or you need to propagate traces in places other than outgoing HTTP requests (e.g. when working with websockers), you can manually set up your application for distributed tracing to work.

## Enabling Automatic Distributed Tracing

To enable distributed tracing for your frontend, add `browserTracingIntegration` to your `Sentry.init()` options as described in the <PlatformLink to="/tracing/instrumentation/automatic-instrumentation/">Automatic Instrumentation</PlatformLink> docs.

> If you want to use distributed tracing without creating spans, set the `tracesSampleRate` option to `0`.

Reviewer: Why would I want to do this? Maybe I completely misunderstand the whole concept of tracing... but how can you trace something without creating spans? How does it look in Sentry?

Author: We have this page that tries to explain the concept of distributed tracing: https://docs.sentry.io/platforms/javascript/guides/react/tracing/trace-propagation/#what-is-distributed-tracing - is that helpful? It's a bit hard to describe/explain this here again without exceeding the scope of it considerably 😬 I can try to put an expandable in there!

Reviewer: Nope sorry, that page doesn't answer why I would want to NOT create spans and still see distributed traces in Sentry. I don't understand.

If you don't want to use the `browserTracingIntegration` integration, or you want to propagate traces in other places than HTTP requests (for example, with websockets), you can manually extract and inject tracing data in your application to connect multiple systems. For this, you must:

- Extract and store incoming tracing information from HTML `<meta>` tags when loading the page.
- Inject tracing information to any outgoing requests.

To learn more about distributed tracing, see our <PlatformLink to="/tracing/trace-propagation/">Distributed Tracing</PlatformLink> docs.

### Extract Tracing Information From HTML `meta` Tags

If you have a server that renders your application's HTML (server-side rendering) and is also running a Sentry SDK, you can connect your backend to your backend via tracing.

To do this, have your server render HTML `<meta>` tags with the Sentry trace information. In your frontend, extract that tracing information when the page is loading and use it to create new transactions connected to that incoming backend trace.

Some Sentry backend SDKs provide a built-in way to inject these `<meta>` tags into rendered HTML. For example:

> - [Python](/platforms/python/distributed-tracing/custom-instrumentation/#inject-tracing-information-into-rendered-html)

Reviewer 2: l/m: I'd add the link to Node docs as well here
```suggestion
- [Node](/platforms/javascript/guides/node/tracing/trace-propagation/custom-instrumentation/#injecting-tracing-information-into-html)
- [Python](/platforms/python/distributed-tracing/custom-instrumentation/#inject-tracing-information-into-rendered-html)
```

> Distributed tracing will be set up automatically if you've <PlatformLink to="/tracing/">Set Up Tracing</PlatformLink>.

Author: I think the idea here was to match the nav label, but I agree it does not really make sense - will change it!

> If you have not enabled tracing, or you need to propagate traces in other places than outgoing HTTP requests, you can manually set up your application for distributed tracing to work.

Reviewer: I'm sorry I have no idea what this sentence is trying to communicate.

Author: I rewrote it a bit, is it clearer now?

Reviewer: So I'm trying to understand this logic.

IF I have not enabled tracing (so I have not set tracesSampleRate)
OR if I need to propagate traces in other places that are NOT HTTP requests (e.g. websocket events, I guess?)
THEN I can do some manual setup to make distributed tracing work

BUT

Why would I want to manually setup distributed tracing, if I have not enabled tracing???
Why would I not enable tracing?
How does distributed tracing work without creating spans?

I'm sorry I must sound so stupid but I think it would be useful to provide some examples here because apart from the OR clause above, I don't know why you would consider the first IF?

Maybe this stuff is covered elsewhere on the docs so maybe it is worth linking to but it just seems so overly complex and counterintuitive to me.

## Enabling Automatic Distributed Tracing

To enable distributed tracing for your frontend, add `browserTracingIntegration` to your `Sentry.init()` options as described in the <PlatformLink to="/tracing/instrumentation/automatic-instrumentation/">Automatic Instrumentation</PlatformLink> docs.

> If you want to use distributed tracing but not tracing, set the `tracesSampleRate` option to `0`.

Reviewer: `If you want to use distributed tracing but not tracing`- this is so confusing. We need to explain this in more detail. Even the SDK option doesn't make much sense, because you've set something to 0 yet it is kind of working in a way because of the integration... it's hurting my head.

Author: Rewrote this a bit, is it clearer now?

Reviewer: It's clearer, yes, but I still have more questions :D added a comment

```javascript
Sentry.init({
  dsn: "___PUBLIC_DSN___",
  integrations: [Sentry.browserTracingIntegration()],

  // If this is 0, tracing is disabled
  // but distributed tracing is not
  tracesSampleRate: 0,
});
```

### Continuing a Trace Manually

> By default, the `browserTracingIntegration` will automatically continue a trace found in a `<meta>` tags - see <PlatformLink to="/tracing/trace-propagation/#automatic-trace-propagation">Automatic Trace Propagation</PlatformLink> for details.

Reviewer 3: ```suggestion
By default, the `browserTracingIntegration` will automatically continue a trace found in a `<meta>` tag - see <PlatformLink to="/tracing/trace-propagation/#automatic-trace-propagation">Automatic Trace Propagation</PlatformLink> for details.
```

> If you want to continue a different trace, for example because you cannot propagate the trace through meta tags but through some different mechanism, you can do this as follows:

Reviewer 3: ```suggestion
If you want to continue a different trace, for example because you cannot propagate the trace through meta tags but through some different mechanism, you can do this:
```

```javascript
const client = Sentry.init({
  dsn: "___PUBLIC_DSN___",
  integrations: [
    Sentry.browserTracingIntegration({
      // Disable default pageload instrumentation
      instrumentPageLoad: false,
    }),
  ],
});

// Start the pageload span manually
const sentryTrace = getSentryTraceStringFromSomewhere();
const baggage = getBaggageStringFromSomewhere();

Sentry.startBrowserTracingPageLoadSpan(
  client,
  {
    name: window.location.pathname,
  },
  { sentryTrace, baggage }
);
```

> ## Configuring Distributed Tracing without `browserTracingIntegration`

Reviewer 3: ```suggestion
## Configuring Distributed Tracing Without `browserTracingIntegration`
```

If you don't want to use the `browserTracingIntegration` integration, you can manually extract and inject tracing data in your application to connect multiple systems. For this, you must:

- Extract and store incoming tracing information from HTML `<meta>` tags when loading the page.
- Inject tracing information to any outgoing requests.

To learn more about distributed tracing, see our <PlatformLink to="/tracing/trace-propagation/">Distributed Tracing</PlatformLink> docs.

### Extract Tracing Information From HTML `meta` Tags

> If you have a server that renders your application's HTML (SSR) and is also running a Sentry SDK, you can connect your backend to your backend via tracing.

Reviewer 3: ```suggestion
If you have a server that renders your application's HTML (server-side rendering) and is also running a Sentry SDK, you can connect your backend to your backend via tracing.
```

Distributed tracing will be set up automatically if you've <PlatformLink to="/tracing/">set up tracing</PlatformLink>.

When tracing is enabled, the Sentry SDK will not only create spans for pageloads and transactions but also propagate trace information to other systems. This allows you to connect multiple systems and see how they interact with each other.

> If you have not enabled tracing, or you need to propagate traces in places other than outgoing HTTP requests (e.g. when working with websockers), you can manually set up your application for distributed tracing to work.

Reviewer: So I'm trying to understand this logic.

IF I have not enabled tracing (so I have not set tracesSampleRate)
OR if I need to propagate traces in other places that are NOT HTTP requests (e.g. websocket events, I guess?)
THEN I can do some manual setup to make distributed tracing work

BUT

Why would I want to manually setup distributed tracing, if I have NOT enabled tracing???
Why would I not enable tracing? Is this because I don't want to propagate trace information to other systems?
How does distributed tracing work without creating spans?

Maybe this stuff is covered elsewhere on the docs so maybe it is worth linking to but it just seems so overly complex and counterintuitive to me. Or maybe the information in the two small paragraphs needs to be combined into one to provide more concrete use cases of each situation.

Author: the only reason to not enable tracing here is to save cost - I do not want to pay for spans. I am not sure how exactly we want to point this out here. Generally we _want_ people to just enable tracing.

Reviewer: Ah ha, I figured this was the case. But the point still stands. How does distributed tracing work without spans if everything is based on spans? It just seems impossible. And maybe we don't even need to go into it, so maybe we don't even call this out, and just talk about tracing in other places that are not HTTP-based?

Author: So I adjusted this some more, after talking with others on the team!

1. I added an expandable that explains what distributed tracing is, for context.
2. I removed the `tracesSampleRate: 0` mentions as that is not even necessary anymore. this should hopefully improve clarity a bit...

Reviewer: This is much clearer now, thank you for taking the time to do this!
